File: project_euler/02/026.py
#!/usr/bin/env python3
'''Copyright [2017] <dmnsn7@gmail.com>'''

D = 1000


def get_recycle_size(num_n):
    remainder = 1
    idx = 1
    visited = {}

    while remainder not in visited:
        visited[remainder], idx = idx, idx + 1

        while remainder != 0 and remainder < num_n:
            remainder *= 10

        remainder %= num_n

    return idx - visited[remainder] if remainder != 0 else 0


def main():
    max_recycle_size, target = 0, 1
    for i in range(1, D):
        if get_recycle_size(i) > max_recycle_size:
            max_recycle_size = get_recycle_size(i)
            target = i

    print(target)


if __name__ == '__main__':
    main()

File: project_euler/02/026.go
// Copyright [2017] <dmnsn7@gmail.com>

package main

import "fmt"

const cD = 1000

func getRecycleSize(n int) int {
	remainder, idx := 1, 1
	visited := map[int]int{}

	for _, exist := visited[remainder]; !exist; {
		visited[remainder] = idx
		idx++
		for remainder != 0 && remainder < n {
			remainder *= 10
		}

		remainder %= n
		_, exist = visited[remainder]
	}

	if remainder != 0 {
		return idx - visited[remainder]
	}
	return 0
}

func main() {
	maxRecycleSize, target := 0, 1
	for i := 1; i < cD; i++ {
		if getRecycleSize(i) > maxRecycleSize {
			maxRecycleSize, target = getRecycleSize(i), i
		}
	}

	fmt.Printf("%d\n", target)
}
